pyvista/core/utilities/helpers.py

```diff
@@ -10,7 +15,9 @@
 from .fileio import from_meshio, is_meshio_mesh
 
 
-def wrap(dataset):
+def wrap(
+    dataset: Union[np.ndarray, _vtk.vtkDataSet, 'Trimesh', 'Mesh']
+) -> Optional[Union['pyvista.DataSet', 'pyvista.pyvista_ndarray']]:
```

Comment:
Can we use an overload to distinguish between `DataSet` return and `pyvista_ndarray` return?

Reply:
I can't for the life of me get overloading to work.

This seems like it should work fine:

```py
@overload
def wrap(dataset: np.ndarray) -> Optional['pyvista.pyvista_ndarray']:
    ...

@overload
def wrap(dataset: Union[_vtk.vtkDataSet, 'Trimesh', 'Mesh']) -> Optional['pyvista.DataSet']:
    ...


def wrap(
    dataset: Union[np.ndarray, _vtk.vtkDataSet, 'Trimesh', 'Mesh']
) -> Union[Optional['pyvista.DataSet'], Optional['pyvista.pyvista_ndarray']]:
```

And yet:
```
pyvista/core/utilities/helpers.py:18: error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]        
```

Reply:
Maybe the `dataset` in the actual implementation should be left untyped?

Also, you could get rid of the `Optional` return type as well with another overload, right?

```python
@overload
def wrap(dataset: None) -> None: ...
```

Reply:
One issue is that you shouldn't type the actual function, this causes the duplicate.  For example:

```py
@overload
def wrap(dataset: np.ndarray) -> Optional['pyvista.pyvista_ndarray']:
    ...

@overload
def wrap(dataset: Union[_vtk.vtkDataSet, 'Trimesh', 'Mesh']) -> Optional['pyvista.DataSet']:
    ...


def wrap(dataset):
    ...
```

Reply:
See a few examples down from here https://peps.python.org/pep-0484/#function-method-overloading

Reply:
I tried @darikg suggestion too and that didn't work.

Also we are missing the possibility that `None` is passed into the function.  I'm surprised that the type checker doesn't complain about this now and this seems like a clue to this problem.

Reply:
Found it. The type of `_vtk.vtkDataSet` is `Any`.  vtk doesn't ship with type hints, although they can be autogenerated.

Reply:
We are not going to be able to use overload unless this can be addressed.

Reply:
Interesting! I gave it a shot. `try_vtk_mypy.py`:

```python
from vtkmodules.vtkCommonDataModel import vtkDataSet

reveal_type(vtkDataSet())
```

At the command line, `mypy try_vtk_mypy.py` says `try_mypy.py:3: note: Revealed type is "Any"`

Autogenerate the stubs: `stubgen --package vtkmodules --output stubs` and `set MYPYPATH=./stubs`, then  `mypy try_vtk_mypy.py` again:

```sh
stubs\vtkmodules\vtkCommonDataModel.pyi:4494: error: invalid syntax  [syntax]
Found 1 error in 1 file (errors prevented further checking)
```

Ok, manually fix some minor syntax errors in `vtkCommonDataModel.pyi`, and after the second and third tries some duplicate function argument names in `vtkCommonCore.pyi` and `vtkCommonMath.pyi`. Then run `mypy try_vtk_mypy.py` again:

```sh
(Thousands of warnings and errors omitted)
try_mypy.py:3: note: Revealed type is "vtkmodules.vtkCommonDataModel.vtkDataSet"
Found 6123 errors in 4 files (checked 1 source file)
```

It works! But there's other 6123 errors to patch up, haha

Reply:
I don't think we can ship typing that requires the user to modify their vtk site-packages.  I think it would require vtk to ship the type stubs.

Reply:
yeah, or manage a separate stubs repo required as a dependency, which seems like a nightmare

Reply:
> yeah, or manage a separate stubs repo required as a dependency, which seems like a nightmare

I think it should be the responsibility of Kitware to do that. Typing for pyvista alone is enough work.